allennlp/training/optimizers.py

```diff
@@ -196,6 +205,149 @@ def default(model_parameters: List) -> "Optimizer":
+            regexes, pg_overrides = parameter_group
+            optimizer_name = pg_overrides.get("optimizer_name", "default")
+            optimizer_name_to_parameter_groups[optimizer_name].append(parameter_group)
+
+        # calculate model parameters for each optimizer
+        optimizer_name_to_model_parameters: Dict[str, List[Tuple[str, torch.nn.Parameter]]] = {
+            optimizer_name: [] for optimizer_name in optimizers.keys()
+        }
+        for model_parameter_tuple in model_parameters:
+            parameter_name, parameter_tensor = model_parameter_tuple
+            for regexes, pg_overrides in parameter_groups:
+                if any(re.search(regex, parameter_name) for regex in regexes):
+                    optimizer_name = pg_overrides.get("optimizer_name", "default")
+                    optimizer_name_to_model_parameters[optimizer_name].append(model_parameter_tuple)
+                    break
+            else:
+                optimizer_name_to_model_parameters["default"].append(model_parameter_tuple)
+
+        # Check to see if an optimizer didn't receive any parameters.
+        for optimizer_name, optimizer_parameters in optimizer_name_to_model_parameters.items():
+            if len(optimizer_parameters[0]) == 0:
+                raise ConfigurationError(
+                    f"Optimizer '{optimizer_name}' did not receive any parameters."
+                    " If you are using `parameter_groups`, please make sure that the regexes you have provided"
+                    " match the desired model parameters, or that the `name` value of this optimizer "
+                    " matches that of the parameter group you are trying to assign to it."
+                    " Alternatively, you can remove this optimizer from the provided `optimizers`"
+                    " if it is not relevant to a particular parameter group."
+                )
+
+        self.optimizers = {
+            optimizer_name: lazy_optimizer.construct(
+                model_parameters=optimizer_name_to_model_parameters[optimizer_name],
+                parameter_groups=optimizer_name_to_parameter_groups[optimizer_name],
+            )
+            for optimizer_name, lazy_optimizer in optimizers.items()
+        }
+
+        # Copy the defaults from the optimizers into the parameter groups, so they are visible in
+        # optimizer.params when optimizer is a RegexOptimizer.
+        parameter_groups = copy.deepcopy(parameter_groups)
+        for parameter_group in parameter_groups:
+            regexes, pg_overrides = parameter_group
+            optimizer_name = pg_overrides.get("optimizer_name", "default")
+            optimizer = self.optimizers[optimizer_name]
+            for key, value in optimizer.defaults.items():
+                if key not in pg_overrides:
+                    pg_overrides[key] = value
+        super().__init__(make_parameter_groups(model_parameters, parameter_groups), {})
```

Comment:
I added this to capture the intent you had where you could still see the correct parameters in the outer optimizer, even if that's not the optimizer that's doing the actual work. I'm not entirely sure we need this though. Thoughts?

Reply:
Yes, I don't think it's completely necessary to be able to view the optimizer options in the outer optimizer groups. The one place I found this to be necessary was when a `scheduler` wants to know the `param_group_field` such as the `"lr"`:

https://github.com/allenai/allennlp/blob/a02f67da036f8f7df4c9bdbd59870686d55bf4c2/allennlp/training/scheduler.py#L36-L39

So when accessing `self.optimizer.param_groups`, the default group only has a 'params' key: 
```
0 dict_keys(['params', 'betas', 'optimizer_name', 'lr', 'eps', 'weight_decay', 'amsgrad'])
1 dict_keys(['params', 'lr', 'optimizer_name', 'betas', 'eps', 'weight_decay', 'amsgrad'])
2 dict_keys(['params', 'optimizer_name', 'lr', 'betas', 'eps', 'weight_decay', 'amsgrad'])
3 dict_keys(['params']) # default group
```

which causes the following error if using a learning rate scheduler (it's line 41 instead of 38 for me as I have some debugging print statements): 
```
# Include a learning_rate_scheduler in the training config:
"learning_rate_scheduler": {
  "type": "linear_with_warmup",
  "warmup_steps": 5000,
},

File "/home/jbarry/allennlp-fork/allennlp/allennlp/training/scheduler.py", line 41, in __init__
    raise KeyError(f"{param_group_field} missing from param_groups[{i}]")
KeyError: 'lr missing from param_groups[3]'
```


The way the code is now, even when copying these optimizer options to each parameter group, the options from the `default` optimizer won't be copied over because the loop operates on [`parameter_groups`](https://github.com/jbrry/allennlp/blob/77ea6a63ca15c92940e1817fa83a28d24c3a11ca/allennlp/training/optimizers.py#L297) and the main `default` group is usually just catching what's left over from `parameter_groups` (unless you want to include some parameters in the `default` optimizer with some non-default options).

Reply:
This seems pretty important. Learning rate schedulers have to work. But how did this work before? I don't think we ever populated the default group this way?

Reply:
I made a change in 63ec4f3817491d748de58b0a2a42116bfd8d2b61 that makes this work again. But I think this scenario would be broken with any other optimizer when using parameter groups. Can you confirm?

Reply:
This now works for me with a `learning_rate_scheduler` and group `3` now looks like (and seems to match the traditional `Optimizer` case):

```
0 dict_keys(['params', 'betas', 'optimizer_name', 'lr', 'eps', 'weight_decay', 'amsgrad'])
1 dict_keys(['params', 'lr', 'optimizer_name', 'betas', 'eps', 'weight_decay', 'amsgrad'])
2 dict_keys(['params', 'optimizer_name', 'lr', 'betas', 'eps', 'weight_decay', 'amsgrad'])
3 dict_keys(['params', 'lr', 'betas', 'eps', 'weight_decay', 'correct_bias'])

Parameter Group 3
    betas: (0.9, 0.999)
    correct_bias: True
    eps: 1e-08
    initial_lr: 0.0004
    lr: 0.0
    weight_decay: 0.01
)

```